sdk/core/Azure.Core/src/Http/RequestUriBuilder.cs

```diff
@@ -192,12 +197,77 @@ public override string ToString()
             else
             {
                 stringBuilder.Append(Uri.EscapeUriString(_pathAndQuery.ToString(0, _queryIndex)));
-                stringBuilder.Append(_pathAndQuery.ToString(_queryIndex, _pathAndQuery.Length - _queryIndex));
+                if (allowedQueryParameters == null)
+                {
+                    stringBuilder.Append(_pathAndQuery.ToString(_queryIndex, _pathAndQuery.Length - _queryIndex));
+                }
+                else
+                {
+                    AppendSanitizedQuery(stringBuilder, allowedQueryParameters);
+                }
             }
 
             return stringBuilder.ToString();
         }
 
+        private void AppendSanitizedQuery(StringBuilder stringBuilder, string[] allowedQueryParameters)
+        {
+            string query = _pathAndQuery.ToString(_queryIndex, _pathAndQuery.Length - _queryIndex);
+            int queryIndex = 1;
+            stringBuilder.Append('?');
+
+            do
+            {
+                int endOfParameterValue = query.IndexOf('&', queryIndex);
+                int endOfParameterName = query.IndexOf('=', queryIndex);
+
+                // Check if we have parameter without value
+                if (endOfParameterValue != -1 &&
+                    (endOfParameterName == -1 || endOfParameterName > endOfParameterValue))
+                {
+                    endOfParameterName = endOfParameterValue;
+                }
+
+                if (endOfParameterName == -1)
+                {
+                    endOfParameterName = query.Length;
+                }
+
+                if (endOfParameterValue == -1)
+                {
+                    endOfParameterValue = query.Length;
+                }
+                else
+                {
+                    // include the separator
+                    endOfParameterValue++;
+                }
+
+                ReadOnlySpan<char> parameterName = query.AsSpan(queryIndex, endOfParameterName - queryIndex);
+
+                bool isAllowed = false;
+                foreach (string name in allowedQueryParameters)
+                {
+                    if (parameterName.Equals(name.AsSpan(), StringComparison.OrdinalIgnoreCase))
```

Comment:
If you get the hashcode of a `ReadOnlySpan<char>`, would it be the same as the corresponding `string`?  I'd like to allow-list all the headers I can get away with and don't want logging to be costlier than it has to be.

Reply:
Unfortunately no, hashcodes are different. I know how to write an optimized version of this but not sure if it's worth it at this moment.

Reply:
I agree - we don't need an optimized version today.